src/ILCompiler.Compiler/src/Compiler/DependencyAnalysis/ReflectionVirtualInvokeMapNode.cs

```diff
@@ -0,0 +1,224 @@
+
+            // Get a list of all methods that have a method body and metadata from the metadata manager.
+            foreach (var mappingEntry in factory.MetadataManager.GetMethodMapping(factory))
+            {
+                MethodDesc method = mappingEntry.Entity;
+
+                // The current format requires us to have an EEType for the owning type. We might want to lift this.
+                if (!factory.MetadataManager.TypeGeneratesEEType(method.OwningType))
+                    continue;
+
+                // We have a method body, we have a metadata token, but we can't get an invoke stub. Bail.
+                if (!factory.MetadataManager.IsReflectionInvokable(method))
+                    continue;
+
+                // Only virtual methods are interesting
+                if (!NeedsVirtualInvokeInfo(method))
+                    continue;
+
+                //
+                // When working with the ProjectN ABI, the entries in the map are based on the definition types. All
+                // instantiations of these definition types will have the same vtable method entries.
+                // On CoreRT, the vtable entries for each instantiated type might not necessarily exist.
+                // Example 1: 
+                //      If there's a call to Foo<string>.Method1 and a call to Foo<int>.Method2, Foo<string> will
+                //      not have Method2 in its vtable and Foo<int> will not have Method1.
+                // Example 2:
+                //      If there's a call to Foo<string>.Method1 and a call to Foo<object>.Method2, given that both
+                //      of these instantiations share the same canonical form, Foo<__Canon> will have both method 
+                //      entries, and therefore Foo<string> and Foo<object> will have both entries too.
+                // For this reason, the entries that we write to the map in CoreRT will be based on the canonical form
+                // of the method's containing type instead of the open type definition.
+                //
+                // Similarly, given that we use the open type definition for ProjectN, the slot numbers ignore dictionary
+                // entries in the vtable, and computing the correct slot number in the presence of dictionary entries is 
+                // done at runtime. When working with the CoreRT ABI, the correct slot numbers will be written to the map,
+                // and no adjustments will be performed at runtime.
+                //
+
+                TypeDesc containingTypeKey;
+
+                if (factory.Target.Abi == TargetAbi.ProjectN)
+                    containingTypeKey = method.OwningType.GetTypeDefinition();
+                else
+                    containingTypeKey = method.OwningType.ConvertToCanonForm(CanonicalFormKind.Specific);
+
+                if (!methodsEmitted.ContainsKey(mappingEntry.MetadataHandle))
```

Comment:
fixed

---
In reply to: [108006180](https://github.com/dotnet/corert/pull/3100#discussion_r108006180) [](ancestors = 108006180)